freedomofpress/securedrop-client
# fix: limit and track processing of `SendReplyJob`s around session and network partitions

* Blocked on #1500

# Description

* Fixes #1420 by clearing the `ApiJobQueue` when it stops, so that unprocessed jobs (e.g. `SendReplyJob`s for pending `DraftReply`s) will not continue to be processed under the session that's just logged out.
* Fixes #1457 and fixes #1493 by having the `SendReplyJob` take "ownership" of a given `DraftReply` by marking it with the current PID, so that both the current and a subsequent Client session can infer whether or not the `SendReplyJob`'s API call could still be in flight.

For more details on why all this extra machinery is necessary to address this pair of corner cases, see especially <https://github.com/freedomofpress/securedrop-client/pull/1486#issuecomment-1136854646> and <https://github.com/freedomofpress/securedrop-client/issues/1457#issuecomment-1123050639>.

# Test Plan

Synthesized from the reproductions of #1420 and #1457:

1. On your test server, watch `/var/log/apache2/journalist-access.log`.
2. Create a new source and submit a message.
3. Log into the Client and wait for sync.
4. Send a reply and wait for it to send.
5. [ ] Observe exactly 1 `"POST /api/v1/sources/<uuid>/replies HTTP/1.1" 201`.
6. Disconnect `sd-whonix` from `sys-firewall`.
7. Send at least `n >= 1` more replies.
8. Quit the Client.
9. Reconnect `sd-whonix` to `sys-firewall`.
10. [ ] Observe exactly 1 `"POST /api/v1/sources/<uuid>/replies HTTP/1.1" 201`.
11. Wait several seconds.
12. Start the Client again.
13. Log in again and wait for sync.
14. [ ] Observe only the first message from step (5) has succeeded...
15. [ ] ...and the rest are marked `Failed to send`.
16. Repeat steps (1)‒(15), but log out (and back in) rather than quitting and restarting the Client.
17. [ ] Observe the same results.

## Bonus: bootstrapping case

1. Delete your `svs.sqlite` and reinitialize it with `alembic upgrade head`.
2. Log into the Client and wait for sync.
3. [ ] Sources and submissions are retrieved and displayed as usual.

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [x] I have written a migration and upgraded a test database based on `main` and confirmed that the migration is [self-contained] and applies cleanly
 - I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - I need help writing a database migration
 - No database schema changes are needed

[self-contained]: https://github.com/freedomofpress/securedrop-client#generating-and-running-database-migrations
